Model based testing

danbornside opened the pull request.

danbornside commented:
same as !2230, but in iohk's repo to enable CI.

JaredCorduan reviewed (approved):
Looks good to me! I think you just need to run ormolu (there is a script in `./scripts/ormolise.sh` that will run it for you if you have nix installed)

nc6 commented on alonzo/impl/src/Cardano/Ledger/Alonzo/Translation.hs:
```diff
@@ -205,7 +205,7 @@ translateTxOut (Shelley.TxOutCompact addr value) =
 
 translatePParams ::
-  AlonzoGenesis -> Shelley.PParams (MaryEra c) -> PParams (AlonzoEra c)
+  AlonzoGenesis -> Shelley.PParams (MaryEra era) -> PParams (AlonzoEra c)
```
This change doesn't make any sense.

danbornside replied:
I guess, `Shelley.PParams` already doesn't make sense to me, it doesn't use it's argument at all; it's not a type family; i don't see why it was as constrained as any of this at all.  Relaxing this constraint allows me to take a fixed `Shelley.PParams` in tests and use it across several eras.

nc6 reviewed (approved):
LGTM. Can you rebase on master?

danbornside committed:
implements model based testing framework.

- duplicate types of many Ledger details, with no crypto details.
- duplicate model implementation of ledger spec
- add a class EraElaborator to apply era-agnostic blockchain model interactions
  to real ledger implementations
- instance of EraElaborator for Alonzo as Shelley eras.
- simple unit test which applies simple transactions using EraElaborator to any
  era.
- model supports basic spending, multi-asset minting, delegation, rewards,
  and plutus scripts, minting, invalid scripts, pool/key deposits.
- end-to-end generated model test, with basic shrinking.
- Some docs to guide adding new eras to model.

JaredCorduan commented:
@danbornside the merge will be unblocked once the commit is signed. if you'd rather not do it, I can sign it for you, just let me know.
